File: python/geny_harness/core/pipeline.py
"""Pipeline engine — executes stages in order with loop control.

Matches geny-executor's Pipeline interface exactly.
"""

from __future__ import annotations

import asyncio
import uuid
from typing import Any, AsyncIterator, Callable, Dict, List, Optional

from geny_harness._native import (
    PipelineConfig,
    PipelineEvent,
    PipelineResult,
    PipelineState,
    StageDescription,
)
from geny_harness.core.stage import Stage
from geny_harness.events.bus import EventBus


class Pipeline:
    """Stage들을 순서대로 실행하는 파이프라인 엔진.

    Execution model:
      Phase A: Input (Stage 1, once)
      Phase B: Agent Loop (Stage 2~13, repeats)
      Phase C: Finalize (Stage 14~16, once)
    """

    # Loop boundary constants
    LOOP_START = 2
    LOOP_END = 13  # inclusive
    FINALIZE_START = 14
    FINALIZE_END = 16  # inclusive
    EVENT_DATA_TRUNCATE = 500  # max chars for event data preview

    # Default names for unregistered stage slots (used in bypass events)
    _DEFAULT_STAGE_NAMES: Dict[int, str] = {
        1: "input", 2: "context", 3: "system", 4: "guard",
        5: "cache", 6: "api", 7: "token", 8: "think",
        9: "parse", 10: "tool", 11: "agent", 12: "evaluate",
        13: "loop", 14: "emit", 15: "memory", 16: "yield",
    }

    def __init__(self, config: Optional[PipelineConfig] = None):
        self._config = config or PipelineConfig()
        self._stages: Dict[int, Stage] = {}
        self._event_bus = EventBus()

    # ── Stage management ──

    def register_stage(self, stage: Stage) -> Pipeline:
        """Register or replace a stage. Supports chaining."""
        self._stages[stage.order] = stage
        return self

    def replace_stage(self, order: int, stage: Stage) -> Pipeline:
        """Replace stage at given order."""
        self._stages[order] = stage
        return self

    def remove_stage(self, order: int) -> Pipeline:
        """Remove stage (that slot will be bypassed)."""
        self._stages.pop(order, None)
        return self

    def get_stage(self, order: int) -> Optional[Stage]:
        """Get registered stage by order."""
        return self._stages.get(order)

    @property
    def stages(self) -> List[Stage]:
        """All registered stages, sorted by order."""
        return sorted(self._stages.values(), key=lambda s: s.order)

    # ── Execution ──

    async def run(self, input: Any, state: Optional[PipelineState] = None) -> PipelineResult:
        """Execute the full pipeline.

        Phase A: Stage 1 (Input) — runs once
        Phase B: Stage 2~13 (Agent Loop) — repeats until loop_decision != "continue"
        Phase C: Stage 14~16 (Finalize) — runs once
        """
        state = self._init_state(state)
        await self._emit("pipeline.start", data={"input": str(input)[: self.EVENT_DATA_TRUNCATE]})

        try:
            # Phase A: Input (stage 1)
            current = await self._run_stage(1, input, state)

            # Phase B: Agent Loop (stages 2~13)
            has_loop_stage = self.LOOP_END in self._stages
            while True:
                for order in range(self.LOOP_START, self.LOOP_END + 1):
                    current = await self._try_run_stage(order, current, state)

                # If no Loop stage is registered, auto-complete after one pass
                if not has_loop_stage and state.loop_decision == "continue":
                    state.loop_decision = "complete"

                if state.loop_decision != "continue":
                    break

                state.iteration += 1
                if state.is_over_iterations:
                    state.loop_decision = "complete"
                    state.completion_signal = "MAX_ITERATIONS"
                    state.add_event(
                        "loop.force_complete",
                        {"reason": "max_iterations", "iteration": state.iteration},
                    )
                    break

            # Phase C: Finalize (stages 14~16)
            for order in range(self.FINALIZE_START, self.FINALIZE_END + 1):
                current = await self._try_run_stage(order, current, state)

            result = PipelineResult.from_state(state)
            await self._emit("pipeline.complete", data={"iterations": state.iteration})
            return result

        except Exception as e:
            await self._emit("pipeline.error", data={"error": str(e)})
            return PipelineResult.error_result(str(e), state)

    async def run_stream(
        self, input: Any, state: Optional[PipelineState] = None
    ) -> AsyncIterator[PipelineEvent]:
        """Streaming mode — yields PipelineEvents in real-time.

        Uses an asyncio.Queue so events emitted mid-stage (e.g. text.delta
        during streaming API calls) are yielded immediately, not buffered
        until stage completion.
        """
        state = self._init_state(state)
        queue: asyncio.Queue = asyncio.Queue()
        _SENTINEL = object()

        # Capture EventBus events (stage.enter/exit/bypass etc.)
        def bus_collector(event: PipelineEvent) -> None:
            queue.put_nowait(event)

        # Capture state.add_event() calls (text.delta, api.request etc.)
        # Since PyO3 PipelineState doesn't support _event_listener,
        # we monkey-patch add_event to also push to queue.
        original_add_event = state.add_event

        def patched_add_event(event_type: str, data=None):
            original_add_event(event_type, data)
            queue.put_nowait(
                PipelineEvent(
                    type=event_type,
                    stage=state.current_stage,
                    iteration=state.iteration,
                    data=data or {},
                )
            )

        state.add_event = patched_add_event  # type: ignore[assignment]

        unsubscribe = self._event_bus.on("*", bus_collector)

        async def _run_pipeline() -> None:
            """Execute pipeline phases, then push sentinel to signal completion."""
            try:
                # Phase A
                current = await self._run_stage(1, input, state)

                # Phase B
                has_loop_stage = self.LOOP_END in self._stages
                while True:
                    for order in range(self.LOOP_START, self.LOOP_END + 1):
                        current = await self._try_run_stage(order, current, state)

                    if not has_loop_stage and state.loop_decision == "continue":
                        state.loop_decision = "complete"

                    if state.loop_decision != "continue":
                        break
                    state.iteration += 1
                    if state.is_over_iterations:
                        state.loop_decision = "complete"
                        state.completion_signal = "MAX_ITERATIONS"
                        state.add_event(
                            "loop.force_complete",
                            {"reason": "max_iterations", "iteration": state.iteration},
                        )
                        break

                # Phase C
                for order in range(self.FINALIZE_START, self.FINALIZE_END + 1):
                    current = await self._try_run_stage(order, current, state)

                queue.put_nowait(
                    PipelineEvent(
                        type="pipeline.complete",
                        data={
                            "result": state.final_text[: self.EVENT_DATA_TRUNCATE],
                            "iterations": state.iteration,
                        },
                    )
                )
            except Exception as e:
                queue.put_nowait(PipelineEvent(type="pipeline.error", data={"error": str(e)}))
            finally:
                queue.put_nowait(_SENTINEL)

        try:
            yield PipelineEvent(
                type="pipeline.start", data={"input": str(input)[: self.EVENT_DATA_TRUNCATE]}
            )

            # Run pipeline in background task so we can yield events as they arrive
            task = asyncio.create_task(_run_pipeline())

            while True:
                event = await queue.get()
                if event is _SENTINEL:
                    break
                yield event

            await task  # propagate any unexpected errors

        except Exception as e:
            yield PipelineEvent(type="pipeline.error", data={"error": str(e)})

        finally:
            state.add_event = original_add_event  # type: ignore[assignment]
            unsubscribe()

    # ── Events ──

    def on(self, event_type: str, handler: Callable) -> Callable:
        """Register event handler. Returns unsubscribe function."""
        return self._event_bus.on(event_type, handler)

    @property
    def event_bus(self) -> EventBus:
        """Access the event bus directly."""
        return self._event_bus

    # ── UI metadata ──

    def describe(self) -> List[StageDescription]:
        """Return pipeline structure for UI rendering."""
        descriptions = []
        for order in range(1, 17):
            stage = self._stages.get(order)
            if stage:
                desc = stage.describe()
                descriptions.append(desc)
            else:
                inactive = StageDescription(
                    name=self._DEFAULT_STAGE_NAMES.get(order, f"stage_{order}"),
                    order=order,
                    category="unregistered",
                )
                inactive.is_active = False
                descriptions.append(inactive)
        return descriptions

    # ── Internal ──

    def _init_state(self, state: Optional[PipelineState]) -> PipelineState:
        """Initialize or apply config to state."""
        state = state or PipelineState()
        if not state.pipeline_id:
            state.pipeline_id = uuid.uuid4().hex[:12]
        self._config.apply_to_state(state)
        return state

    async def _try_run_stage(self, order: int, current: Any, state: PipelineState) -> Any:
        """Run a stage if it exists and should not be bypassed."""
        stage = self._stages.get(order)
        if stage is None:
            name = self._DEFAULT_STAGE_NAMES.get(order, f"stage_{order}")
            await self._emit("stage.bypass", stage=name, iteration=state.iteration)
            return current
        if stage.should_bypass(state):
            await self._emit("stage.bypass", stage=stage.name, iteration=state.iteration)
            return current
        return await self._run_stage(order, current, state)

    async def _run_stage(self, order: int, input: Any, state: PipelineState) -> Any:
        """Execute a single stage with lifecycle hooks."""
        stage = self._stages.get(order)
        if stage is None:
            return input

        state.current_stage = stage.name
        # NOTE: stage_history is a PyO3 list copy, so we use add_event for tracking instead
        state.add_event("stage.history", {"stage": stage.name})
        await self._emit("stage.enter", stage=stage.name, iteration=state.iteration)

        await stage.on_enter(state)
        try:
            result = await stage.execute(input, state)
            await stage.on_exit(result, state)
            await self._emit("stage.exit", stage=stage.name, iteration=state.iteration)
            return result
        except Exception as e:
            await self._emit(
                "stage.error",
                stage=stage.name,
                iteration=state.iteration,
                data={"error": str(e)},
            )
            recovery = await stage.on_error(e, state)
            if recovery is not None:
                return recovery
            raise

    async def _emit(self, event_type: str, **kwargs: Any) -> None:
        """Emit a pipeline event."""
        event = PipelineEvent(type=event_type, **kwargs)
        await self._event_bus.emit(event)

File: crates/geny-harness-core/src/core/pipeline.rs
//! Pipeline engine — executes stages in order with loop control.

use std::collections::HashMap;
use std::sync::Arc;

use serde_json::Value;
use tokio::sync::mpsc;

use crate::core::config::PipelineConfig;
use crate::core::errors::StageError;
use crate::core::result::PipelineResult;
use crate::core::stage::{Stage, StageDescription};
use crate::core::state::PipelineState;
use crate::events::bus::{EventBus, EventHandler};
use crate::events::types::PipelineEvent;

/// Pipeline engine — Stage들을 순서대로 실행하는 파이프라인 엔진.
///
/// Execution model:
///   Phase A: Input (Stage 1, once)
///   Phase B: Agent Loop (Stage 2~13, repeats)
///   Phase C: Finalize (Stage 14~16, once)
pub struct Pipeline {
    config: PipelineConfig,
    stages: HashMap<u32, Box<dyn Stage>>,
    event_bus: Arc<EventBus>,
}

impl Pipeline {
    // Loop boundary constants
    pub const LOOP_START: u32 = 2;
    pub const LOOP_END: u32 = 13; // inclusive
    pub const FINALIZE_START: u32 = 14;
    pub const FINALIZE_END: u32 = 16; // inclusive
    pub const EVENT_DATA_TRUNCATE: usize = 500; // max chars for event data preview

    /// Default names for unregistered stage slots (used in bypass events).
    fn default_stage_name(order: u32) -> &'static str {
        match order {
            1 => "input",
            2 => "context",
            3 => "system",
            4 => "guard",
            5 => "cache",
            6 => "api",
            7 => "token",
            8 => "think",
            9 => "parse",
            10 => "tool",
            11 => "agent",
            12 => "evaluate",
            13 => "loop",
            14 => "emit",
            15 => "memory",
            16 => "yield",
            _ => "unknown",
        }
    }

    pub fn new(config: Option<PipelineConfig>) -> Self {
        Self {
            config: config.unwrap_or_default(),
            stages: HashMap::new(),
            event_bus: Arc::new(EventBus::new()),
        }
    }

    // ── Stage management ──

    /// Register or replace a stage. Supports chaining.
    pub fn register_stage(&mut self, stage: Box<dyn Stage>) -> &mut Self {
        let order = stage.order();
        self.stages.insert(order, stage);
        self
    }

    /// Replace stage at given order.
    pub fn replace_stage(&mut self, order: u32, stage: Box<dyn Stage>) -> &mut Self {
        self.stages.insert(order, stage);
        self
    }

    /// Remove stage (that slot will be bypassed).
    pub fn remove_stage(&mut self, order: u32) -> &mut Self {
        self.stages.remove(&order);
        self
    }

    /// Get registered stage by order.
    pub fn get_stage(&self, order: u32) -> Option<&dyn Stage> {
        self.stages.get(&order).map(|s| s.as_ref())
    }

    /// All registered stages, sorted by order.
    pub fn stages(&self) -> Vec<&dyn Stage> {
        let mut stages: Vec<_> = self.stages.values().map(|s| s.as_ref()).collect();
        stages.sort_by_key(|s| s.order());
        stages
    }

    // ── Execution ──

    /// Execute the full pipeline.
    ///
    /// Phase A: Stage 1 (Input) — runs once
    /// Phase B: Stage 2~13 (Agent Loop) — repeats until loop_decision != "continue"
    /// Phase C: Stage 14~16 (Finalize) — runs once
    pub async fn run(&self, input: Value, state: Option<PipelineState>) -> PipelineResult {
        let mut state = self.init_state(state);
        let input_preview = truncate_str(&input.to_string(), Self::EVENT_DATA_TRUNCATE);
        self.emit_event("pipeline.start", |e| {
            e.with_data_value("input", Value::String(input_preview))
        })
        .await;

        match self.run_phases(input, &mut state).await {
            Ok(()) => {
                let result = PipelineResult::from_state(&state);
                self.emit_event("pipeline.complete", |e| {
                    e.with_data_value("iterations", Value::Number(state.iteration.into()))
                })
                .await;
                result
            }
            Err(e) => {
                self.emit_event("pipeline.error", |ev| {
                    ev.with_data_value("error", Value::String(e.to_string()))
                })
                .await;
                PipelineResult::error_result(&e.to_string(), Some(&state))
            }
        }
    }

    /// Internal: execute all three phases.
    async fn run_phases(
        &self,
        input: Value,
        state: &mut PipelineState,
    ) -> Result<(), Box<dyn std::error::Error + Send + Sync>> {
        // Phase A: Input (stage 1)
        let mut current = self.run_stage(1, input, state).await?;

        // Phase B: Agent Loop (stages 2~13)
        let has_loop_stage = self.stages.contains_key(&Self::LOOP_END);
        loop {
            for order in Self::LOOP_START..=Self::LOOP_END {
                current = self.try_run_stage(order, current, state).await?;
            }

            // If no Loop stage is registered, auto-complete after one pass
            if !has_loop_stage && state.loop_decision == "continue" {
                state.loop_decision = "complete".to_string();
            }

            if state.loop_decision != "continue" {
                break;
            }

            state.iteration += 1;
            if state.is_over_iterations() {
                state.loop_decision = "complete".to_string();
                state.completion_signal = Some("MAX_ITERATIONS".to_string());
                state.add_event(
                    "loop.force_complete",
                    Some(serde_json::json!({
                        "reason": "max_iterations",
                        "iteration": state.iteration,
                    })),
                );
                break;
            }
        }

        // Phase C: Finalize (stages 14~16)
        for order in Self::FINALIZE_START..=Self::FINALIZE_END {
            current = self.try_run_stage(order, current, state).await?;
        }
        let _ = current;

        Ok(())
    }

    /// Streaming mode — returns a channel receiver for PipelineEvents.
    pub async fn run_stream(
        &self,
        input: Value,
        state: Option<PipelineState>,
    ) -> mpsc::Receiver<PipelineEvent> {
        let (tx, rx) = mpsc::channel(256);

        let mut state = self.init_state(state);

        // Send initial event
        let input_preview = truncate_str(&input.to_string(), Self::EVENT_DATA_TRUNCATE);
        let _ = tx
            .send(
                PipelineEvent::new("pipeline.start")
                    .with_data_value("input", Value::String(input_preview)),
            )
            .await;

        // Capture EventBus events
        let tx_bus = tx.clone();
        let _unsub = self.event_bus.on_sync("*", move |event| {
            let _ = tx_bus.try_send(event.clone());
        });

        // Capture state.add_event() calls
        let tx_state = tx.clone();
        state.event_listener = Some(Box::new(move |event_dict| {
            let event = PipelineEvent {
                event_type: event_dict
                    .get("type")
                    .and_then(|v| v.as_str())
                    .unwrap_or("")
                    .to_string(),
                stage: event_dict
                    .get("stage")
                    .and_then(|v| v.as_str())
                    .unwrap_or("")
                    .to_string(),
                iteration: event_dict
                    .get("iteration")
                    .and_then(|v| v.as_u64())
                    .unwrap_or(0) as u32,
                timestamp: event_dict
                    .get("timestamp")
                    .and_then(|v| v.as_str())
                    .unwrap_or("")
                    .to_string(),
                data: event_dict
                    .get("data")
                    .and_then(|v| v.as_object())
                    .map(|m| m.iter().map(|(k, v)| (k.clone(), v.clone())).collect())
                    .unwrap_or_default(),
            };
            let _ = tx_state.try_send(event);
        }));

        // Run pipeline in background task
        let tx_final = tx;
        // We need to move self's stage references — clone what we need
        // Since we can't move self, we'll need to restructure this
        // For now, run phases and send completion
        tokio::spawn(async move {
            match self_run_phases_standalone(input, &mut state).await {
                Ok(()) => {
                    let _ =
                        tx_final
                            .send(PipelineEvent::new("pipeline.complete").with_data_value(
                                "iterations",
                                Value::Number(state.iteration.into()),
                            ))
                            .await;
                }
                Err(e) => {
                    let _ = tx_final
                        .send(
                            PipelineEvent::new("pipeline.error")
                                .with_data_value("error", Value::String(e.to_string())),
                        )
                        .await;
                }
            }
        });

        rx
    }

    // ── Events ──

    /// Register event handler. Returns unsubscribe function.
    pub fn on(&self, event_type: &str, handler: EventHandler) -> Box<dyn Fn() + Send + Sync> {
        self.event_bus.on(event_type, handler)
    }

    /// Access the event bus directly.
    pub fn event_bus(&self) -> &EventBus {
        &self.event_bus
    }

    // ── UI metadata ──

    /// Return pipeline structure for UI rendering.
    pub fn describe(&self) -> Vec<StageDescription> {
        (1..=16)
            .map(|order| {
                if let Some(stage) = self.stages.get(&order) {
                    stage.describe()
                } else {
                    StageDescription::inactive(
                        Self::default_stage_name(order),
                        order,
                        "unregistered",
                    )
                }
            })
            .collect()
    }

    // ── Internal ──

    /// Initialize or apply config to state.
    fn init_state(&self, state: Option<PipelineState>) -> PipelineState {
        let mut state = state.unwrap_or_default();
        if state.pipeline_id.is_empty() {
            state.pipeline_id = uuid::Uuid::new_v4().to_string()[..12].to_string();
        }
        self.config.apply_to_state(&mut state);
        state
    }

    /// Run a stage if it exists and should not be bypassed.
    async fn try_run_stage(
        &self,
        order: u32,
        current: Value,
        state: &mut PipelineState,
    ) -> Result<Value, Box<dyn std::error::Error + Send + Sync>> {
        let stage = match self.stages.get(&order) {
            None => {
                let name = Self::default_stage_name(order);
                self.emit_event("stage.bypass", |e| {
                    e.with_stage(name).with_iteration(state.iteration)
                })
                .await;
                return Ok(current);
            }
            Some(s) => s,
        };

        if stage.should_bypass(state) {
            self.emit_event("stage.bypass", |e| {
                e.with_stage(stage.name()).with_iteration(state.iteration)
            })
            .await;
            return Ok(current);
        }

        self.run_stage(order, current, state).await
    }

    /// Execute a single stage with lifecycle hooks.
    async fn run_stage(
        &self,
        order: u32,
        input: Value,
        state: &mut PipelineState,
    ) -> Result<Value, Box<dyn std::error::Error + Send + Sync>> {
        let stage = match self.stages.get(&order) {
            Some(s) => s,
            None => return Ok(input),
        };

        state.current_stage = stage.name().to_string();
        state.stage_history.push(stage.name().to_string());
        self.emit_event("stage.enter", |e| {
            e.with_stage(stage.name()).with_iteration(state.iteration)
        })
        .await;

        stage.on_enter(state).await?;

        match stage.execute(input, state).await {
            Ok(result) => {
                stage.on_exit(&result, state).await?;
                self.emit_event("stage.exit", |e| {
                    e.with_stage(stage.name()).with_iteration(state.iteration)
                })
                .await;
                Ok(result)
            }
            Err(e) => {
                self.emit_event("stage.error", |ev| {
                    ev.with_stage(stage.name())
                        .with_iteration(state.iteration)
                        .with_data_value("error", Value::String(e.to_string()))
                })
                .await;

                if let Some(recovery) = stage.on_error(&e, state).await {
                    return Ok(recovery);
                }

                Err(Box::new(StageError::with_cause(
                    e.message.clone(),
                    stage.name(),
                    order,
                    e,
                )))
            }
        }
    }

    /// Emit a pipeline event with builder.
    async fn emit_event<F>(&self, event_type: &str, builder: F)
    where
        F: FnOnce(PipelineEvent) -> PipelineEvent,
    {
        let event = builder(PipelineEvent::new(event_type));
        self.event_bus.emit(&event).await;
    }
}

// Standalone helper for run_stream — temporary until Arc<Pipeline> refactor
async fn self_run_phases_standalone(
    _input: Value,
    _state: &mut PipelineState,
) -> Result<(), Box<dyn std::error::Error + Send + Sync>> {
    // TODO: This needs proper implementation with Arc<Pipeline>
    // For now, run_stream is a structural placeholder
    Ok(())
}

fn truncate_str(s: &str, max_len: usize) -> String {
    if s.len() <= max_len {
        s.to_string()
    } else {
        s[..max_len].to_string()
    }
}

impl std::fmt::Debug for Pipeline {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        f.debug_struct("Pipeline")
            .field("config_name", &self.config.name)
            .field("stage_count", &self.stages.len())
            .field("registered_orders", &{
                let mut orders: Vec<u32> = self.stages.keys().copied().collect();
                orders.sort();
                orders
            })
            .finish()
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::core::stage::Stage as StageTrait;

    struct MockStage {
        name: String,
        order: u32,
        output: Value,
    }

    #[async_trait::async_trait]
    impl StageTrait for MockStage {
        fn name(&self) -> &str {
            &self.name
        }
        fn order(&self) -> u32 {
            self.order
        }
        async fn execute(
            &self,
            _input: Value,
            state: &mut PipelineState,
        ) -> Result<Value, StageError> {
            if self.order == 16 {
                state.final_text = self.output.as_str().unwrap_or("").to_string();
            }
            if self.order == 13 {
                state.loop_decision = "complete".to_string();
            }
            Ok(self.output.clone())
        }
    }

    fn make_mock(name: &str, order: u32, output: &str) -> Box<dyn StageTrait> {
        Box::new(MockStage {
            name: name.to_string(),
            order,
            output: Value::String(output.to_string()),
        })
    }

    #[tokio::test]
    async fn test_minimal_pipeline() {
        let mut pipeline = Pipeline::new(None);
        pipeline.register_stage(make_mock("input", 1, "normalized"));
        pipeline.register_stage(make_mock("api", 6, "response"));
        pipeline.register_stage(make_mock("parse", 9, "parsed"));
        pipeline.register_stage(make_mock("yield", 16, "final"));

        let result = pipeline.run(Value::String("hello".to_string()), None).await;

        assert!(result.success);
        assert_eq!(result.text, "final");
    }

    #[tokio::test]
    async fn test_register_replace_remove() {
        let mut pipeline = Pipeline::new(None);
        pipeline.register_stage(make_mock("input", 1, "v1"));
        assert!(pipeline.get_stage(1).is_some());

        pipeline.replace_stage(1, make_mock("input_v2", 1, "v2"));
        assert_eq!(pipeline.get_stage(1).unwrap().name(), "input_v2");

        pipeline.remove_stage(1);
        assert!(pipeline.get_stage(1).is_none());
    }

    #[tokio::test]
    async fn test_describe() {
        let mut pipeline = Pipeline::new(None);
        pipeline.register_stage(make_mock("input", 1, ""));
        pipeline.register_stage(make_mock("api", 6, ""));

        let descriptions = pipeline.describe();
        assert_eq!(descriptions.len(), 16);
        assert!(descriptions[0].is_active); // stage 1 registered
        assert!(!descriptions[1].is_active); // stage 2 not registered
        assert!(descriptions[5].is_active); // stage 6 registered
    }
}
